File: unpruuf/ios/UnpruufCore/Sources/UnpruufCore/MessagePayload.swift
import Foundation

/// Binary layout of the plaintext a Double Ratchet message carries: `[type: 1 byte][payload]`.
/// Lives *inside* the ratchet ciphertext — see ``RatchetFrame`` for the outer chunking/framing
/// that wraps the ciphertext itself. Direct port of the Android app's `MessagePayload.kt`.
public enum MessagePayload {
    private static let typeText: UInt8 = 0
    private static let typeFile: UInt8 = 1
    private static let typeImage: UInt8 = 2

    public enum Content {
        case text(String)
        case attachment(name: String, bytes: Data, isImage: Bool)
    }

    public static func encodeText(_ text: String) -> Data {
        var out = Data([typeText])
        out.append(text.data(using: .utf8) ?? Data())
        return out
    }

    public static func encodeAttachment(name: String, bytes: Data, isImage: Bool) throws -> Data {
        guard let nameBytes = name.data(using: .utf8), nameBytes.count <= 0xFFFF else {
            throw MessagePayloadError.nameTooLong
        }
        var out = Data(capacity: 1 + 2 + nameBytes.count + bytes.count)
        out.append(isImage ? typeImage : typeFile)
        out.append(UInt8((nameBytes.count >> 8) & 0xFF))
        out.append(UInt8(nameBytes.count & 0xFF))
        out.append(nameBytes)
        out.append(bytes)
        return out
    }

    public static func decode(_ raw: Data) -> Content? {
        guard !raw.isEmpty else { return nil }
        let base = raw.startIndex
        switch raw[base] {
        case typeText:
            let textData = raw.subdata(in: (base + 1)..<raw.endIndex)
            return .text(String(data: textData, encoding: .utf8) ?? "")
        case typeFile, typeImage:
            guard raw.count >= 3 else { return nil }
            let nameLen = (Int(raw[base + 1]) << 8) | Int(raw[base + 2])
            guard raw.count >= 3 + nameLen else { return nil }
            let nameData = raw.subdata(in: (base + 3)..<(base + 3 + nameLen))
            guard let name = String(data: nameData, encoding: .utf8) else { return nil }
            let bytes = raw.subdata(in: (base + 3 + nameLen)..<raw.endIndex)
            return .attachment(name: name, bytes: bytes, isImage: raw[base] == typeImage)
        default:
            return nil
        }
    }
}

public enum MessagePayloadError: Error {
    case nameTooLong
}

File: unpruuf/app/src/main/java/com/nexonai/unpruuf/domain/network/MessagePayload.kt
package com.nexonai.unpruuf.domain.network

/**
 * Binary layout of the plaintext a Double Ratchet message carries: `[type: 1 byte][payload]`.
 * Lives *inside* the ratchet ciphertext — see [P2PNetworkManager] for the outer chunking/framing
 * that wraps the ciphertext itself.
 */
object MessagePayload {

    private const val TYPE_TEXT: Byte = 0
    private const val TYPE_FILE: Byte = 1
    private const val TYPE_IMAGE: Byte = 2

    sealed class Content {
        data class Text(val text: String) : Content()
        data class Attachment(val name: String, val bytes: ByteArray, val isImage: Boolean) : Content()
    }

    fun encodeText(text: String): ByteArray = byteArrayOf(TYPE_TEXT) + text.toByteArray(Charsets.UTF_8)

    fun encodeAttachment(name: String, bytes: ByteArray, isImage: Boolean): ByteArray {
        val nameBytes = name.toByteArray(Charsets.UTF_8)
        require(nameBytes.size <= 0xFFFF) { "file name too long" }
        val out = ByteArray(1 + 2 + nameBytes.size + bytes.size)
        out[0] = if (isImage) TYPE_IMAGE else TYPE_FILE
        out[1] = (nameBytes.size ushr 8).toByte()
        out[2] = nameBytes.size.toByte()
        System.arraycopy(nameBytes, 0, out, 3, nameBytes.size)
        System.arraycopy(bytes, 0, out, 3 + nameBytes.size, bytes.size)
        return out
    }

    fun decode(raw: ByteArray): Content? {
        if (raw.isEmpty()) return null
        return when (raw[0]) {
            TYPE_TEXT -> Content.Text(String(raw, 1, raw.size - 1, Charsets.UTF_8))
            TYPE_FILE, TYPE_IMAGE -> {
                if (raw.size < 3) return null
                val nameLen = ((raw[1].toInt() and 0xFF) shl 8) or (raw[2].toInt() and 0xFF)
                if (raw.size < 3 + nameLen) return null
                val name = String(raw, 3, nameLen, Charsets.UTF_8)
                val bytes = raw.copyOfRange(3 + nameLen, raw.size)
                Content.Attachment(name, bytes, isImage = raw[0] == TYPE_IMAGE)
            }
            else -> null
        }
    }
}
